File: solutions/python/crypto-square/1/crypto_square.py
from math import ceil, floor, sqrt


def cipher_text(plain_text):
    clean_text = "".join(s.lower() for s in plain_text if s.isalnum())

    c = ceil(sqrt(len(clean_text)))
    r = floor(sqrt(len(clean_text)))
    if r * c < len(clean_text):
        r = c

    return "".join(
        (clean_text[j] if j < len(clean_text) else " ")
        + (" " if i < c - 1 and j + c >= r * c else "")
        for i in range(c)
        for j in range(i, r * c, c)
    )

File: solutions/c/crypto-square/1/crypto_square.c
#include "crypto_square.h"

#include <ctype.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>

char *ciphertext(const char *input) {
    char *clean = calloc(strlen(input) + 1, sizeof(char));
    int clean_idx = 0;

    for (; *input; input++)
        if (isalnum(*input)) clean[clean_idx++] = tolower(*input);

    int col = ceil(sqrt(clean_idx));
    int row = floor(sqrt(clean_idx));
    if (row * col < clean_idx) {
        row = col;
    }

    char *final = calloc(row * col + col, sizeof(char));
    int final_idx = 0;

    for (int i = 0; i < col; i++) {
        for (int j = i; j < row * col; j += col) {
            if (j >= clean_idx)
                final[final_idx++] = ' ';
            else
                final[final_idx++] = clean[j];
        }
        if (i < col - 1) final[final_idx++] = ' ';
    }

    free(clean);
    return final;
}
